rubullet/src/client.rs

```diff
@@ -0,0 +1,3254 @@
+                distance,
+                yaw,
+                pitch,
+                roll,
+                up_axis_index,
+                view_matrix.as_mut_ptr(),
+            );
+            view_matrix
+        }
+    }
+    /// computes the projection matrix which can be used together with the view matrix to generate
+    /// camera images within the simulation
+    ///
+    /// # Arguments
+    /// * `left` - left screen (canvas) coordinate
+    /// * `right` - right screen (canvas) coordinate
+    /// * `bottom` - bottom screen (canvas) coordinate
+    /// * `top` - top screen (canvas) coordinate
+    /// * `near` - near plane distance
+    /// * `far` - far plane distance
+    ///
+    /// # See also
+    /// * [compute_view_matrix](`Self::compute_view_matrix`)
+    /// * [compute_view_matrix_from_yaw_pitch_roll](`Self::compute_view_matrix_from_yaw_pitch_roll`)
+    /// * [compute_projection_matrix_fov](`Self::compute_projection_matrix_fov`)
+    /// * [get_camera_image](`Self::get_camera_image`)
+    pub fn compute_projection_matrix(
+        left: f32,
+        right: f32,
+        bottom: f32,
+        top: f32,
+        near_val: f32,
+        far_val: f32,
+    ) -> [f32; 16] {
+        let mut projection_matrix = [0_f32; 16];
+        unsafe {
+            ffi::b3ComputeProjectionMatrix(
+                left,
+                right,
+                bottom,
+                top,
+                near_val,
+                far_val,
+                projection_matrix.as_mut_ptr(),
+            );
+            projection_matrix
+        }
+    }
+    /// computes the projection matrix which can be used together with the view matrix to generate
+    /// camera images within the simulation
+    ///
+    /// # Arguments
+    /// * `fov` - left screen (canvas) coordinate
+    /// * `aspect` - right screen (canvas) coordinate
+    /// * `near_val` - near plane distance
+    /// * `far_val` - far plane distance
+    ///
+    /// # See also
+    /// * [compute_view_matrix](`Self::compute_view_matrix`)
+    /// * [compute_view_matrix_from_yaw_pitch_roll](`Self::compute_view_matrix_from_yaw_pitch_roll`)
+    /// * [compute_projection_matrix](`Self::compute_projection_matrix`)
+    /// * [get_camera_image](`Self::get_camera_image`)
+    pub fn compute_projection_matrix_fov(
+        fov: f32,
+        aspect: f32,
+        near_val: f32,
+        far_val: f32,
+    ) -> [f32; 16] {
+        let mut projection_matrix = [0_f32; 16];
+        unsafe {
+            ffi::b3ComputeProjectionMatrixFOV(
+                fov,
+                aspect,
+                near_val,
+                far_val,
+                projection_matrix.as_mut_ptr(),
+            );
+            projection_matrix
+        }
+    }
+    /// returns an RGBA, depth and segmentation images.
+    ///
+    /// # Note
+    /// Depth and segmentation images are currently not really images as the image crate does not
+    /// properly support these types yet.
```

Comment:
I also thought about that. I am even considering removing the image crate entirely. There is no way to access video devices in the image crate and there is no other crate that can open a video device and outputs an image type from the image crate. The only way to do this is using the opencv crate, but it seems like it is not yet in a reliable state. A good idea would be leaving it up to the user to decide what he wants to do with this data. The problem with Matrices is that they are column-major and images are row-major so in most cases the data would be transformed twice. Maybe we should create our own image type? something like a vec with width, height and number of channels.

Reply:
> The only way to do this is using the opencv crate, but it seems like it is not yet in a reliable state.

I would be against adding OpenCV as a dependency for this (though could maybe convinced to make it a non-default optional dependency).

>  I am even considering removing the image crate entirely. ... Maybe we should create our own image type? something like a vec with width, height and number of channels.

What is the advantage of our own image type over including a [non-default feature image][1] crate? If all we're looking for is an image type, that one is commonly used and would have minimal impact on compile times.

> There is no way to access video devices in the image crate and there is no other crate that can open a video device and outputs an image type from the image crate.

I'm not sure what you're arguing for with this.  If you're talking about opening a camera or something, that's out of the scope of this project. If you're talking about displaying the rendered image on-screen, then that's actually pretty simple with crates like [minifb][2] and isn't really in the scope of this crate. Or am I totally not understanding?

[1]: https://github.com/image-rs/image/blob/master/Cargo.toml#L50
[2]: https://crates.io/crates/minifb

Reply:
> What is the advantage of our own image type over including a non-default feature image crate? If all we're looking for is an  image type, that one is commonly used and would have minimal impact on compile times.

That was what I was looking for. The problem I had with the crate is that we can only use it for specifying the type. I was not able to create really images with the depth and segmentation images, just image buffers (maybe you have an idea on that?). And for that little it seemed to have too much dependencies and slowed down the compilation. However, I played around with it for a while and yes the crate is not so heavy as i have thought. So I would say we should keep it and do a public export of the image types.

Of course RuBullet should not open video devices or something like that. But I expect the image processing library I use could do this and that it also supports continuously drawing of the images from my video steam (like cv::imgshow and cv::waitKey(1)). Just saving the image to disk is not enough for analyzing a video feed. And with that in mind I would choose my image processing library.

In c++ and in python you would obviously choose opencv. However, for rust it is different. On one hand, you have a very widely used image crate which supports the imageproc crate, but it is right now not possible to use it with real hardware from within these libraries and I think they consider it as out of scope. On the other hand, you have the fully featured opencv crate, which seems like it is far from being stable, badly documented and needs a working opencv installation which would make it pain to use as a dependency in any library. 

So for somebody who wants to write a image-based robotics application on real hardware, it is right now not clear how he would do that. And the simulation should not make it hard for the user to adapt his vision pipeline. So I do not want to be locked-in on this decision. I think staying with image crate and hoping that they will make a nice interface for floating point types is a good decision. And if they decide to support more and more file extensions we could just remove all of those features and just enable the "png" feature. The user should then be able to easily convert it to an image processing library of his choice.

Reply:
I might have a more detailed response to this in a bit (need to prepare for my recitation soon).

I have viewed this crate's responsibility regarding images as just providing a type the gives more structure to the bytes than just a chunk of memory. How the consumer wants to use those bytes is entirely up to them. The `ImageBuffer::as_raw` function would return a raw chunk of memory (in the form `&Vec<f32>`) that the user could then handle as they want (feeding it into a neuralnet, for example). By using the `image` crate, the de facto standard,  we increase the chances that the type we return already as an `into` function for the type they are using for their processing.

In the case of OpenCV, it would just be `let mat = Mat::from_slice(image.as_raw());`.